### DISCUSSION TITLE
How can I kill all BrowserSubProcesses for particular browser instance?

### DISCUSSION THREAD

**Discussion by assopri** [cefsharp/CefSharp#3567] [Q&A]
In my app I can create and dispose winforms cefsharp browsers for many times, so I need to close all browser BrowserSubProcesses (related to particular browser) right after I dispose it. So they don`t accumulate during app work. One more detail - I cannot kill subprocesses for all browsers at once, because while one browser should be stopped, others can continue working. How can I achieve that?

Currently, when I create browser - 2 new BrowserSubProcess . When I load smth - the 3rd one appears . When I dispose browser  - just one of them is removed (another 2 remain). I thought it was related to form (hosting browser), however it works in the same way for Offscreen version.

My OS is Windows7 64x , VS 2019 installed. I use Winforms Cefsharp version, .NET 4.5.2 . Cef version is 90.6.50.0

**Comment by amaitland** [comment]
> so I need to close all browser BrowserSubProcesses (related to particular browser) right after I dispose it. So they don`t accumulate during app work. 

They are likely already being closed correctly. It's likely what you are seeing is the spare render process, more details at https://www.magpcss.org/ceforum/viewtopic.php?f=6&t=18214

**Reply by assopri** [reply]
I see, however I still get the following error message, when my app actually closes :
![image](https://user-images.githubusercontent.com/73235464/118629387-341afb80-b7d6-11eb-8935-ab2fecd310db.png) 

And as I can create tons of browsers during app work, I cannot afford myself to accumulate these processes in memory (so they need to be shut down immediately after each particular browser closes).

Any ideas what is the best solution in my situation?

**Reply by amaitland** [reply]
> I see, however I still get the following error message, when my app actually closes :

This has nothing to do with the extra browser process.

As the error message says `Cef.Shutdown` is being called on a different thread to `Cef.Initialize`. They must be called on the same thread.

Are you manually calling `Cef.Shutdown()`?

**Reply by amaitland** [reply]
Actually it looks like you called `Cef.Initialize` on a thread other than the main `WinForms UI` thread.

**Reply by assopri** [reply]
Actually, I do not call Cef.Initialize at all. All browsers in my app are required to have isolated cache. So I use code like this to init browser:
`public override void CreateBrowserObject(BrowserSettings browserSetting)
        {

            Browser = new ChromiumWebBrowser();
            (Browser as ChromiumWebBrowser).BrowserSettings = browserSetting;
          
            RequestContextSettings requestContextSettings = new RequestContextSettings();

            (Browser as ChromiumWebBrowser).RequestContext = new RequestContext(requestContextSettings);
}`

And this code is called on a thread other than the main WinForms UI thread.

**Reply by amaitland** [reply]
> Actually, I do not call Cef.Initialize at all. 

In that case the first time you create a  new `ChromiumWebBrowser` instance `Cef.Initialze()` will be called. See https://github.com/cefsharp/CefSharp/wiki/General-Usage#initialize-and-shutdown for more detail

> And this code is called on a thread other than the main WinForms UI thread.

You will either need to call this code on the `WinForms UI` thread or explicitly call `Cef.Initialze(new CefSettings())`
